lib/types.nix

```
@@ -137,47 +137,63 @@ rec {
-            Lua code inserted at the start of the plugin's configuration.
-            This is the same as using `lib.nixvim.utils.mkBeforeSection` when defining `content`.
-          '';
-        };
-        post = lib.mkOption {
-          type = with types; nullOr lines;
-          default = null;
-          description = ''
-            Lua code inserted at the end of the plugin's configuration.
-            This is the same as using `lib.nixvim.utils.mkAfterSection` when defining `content`.
-          '';
-        };
-        content = lib.mkOption {
-          type = types.lines;
-          default = "";
-          description = ''
-            Configuration of the plugin.
-
-            If `pre` and/or `post` are non-null, they will be merged using the order priorities
-            ${toString (mkBeforeSection null).priority} and ${toString (mkBeforeSection null).priority}
-            respectively.
-          '';
+
+      commonModule = {
+        options = {
+          pre = lib.mkOption {
+            type = with types; nullOr lines;
+            default = null;
+            description =
+              ''
+                Lua code inserted at the start of the plugin's configuration.
+              ''
+              + lib.optionalString hasContent ''
+                This is the same as using `lib.nixvim.utils.mkBeforeSection` when defining `content`.
+              '';
+          };
+          post = lib.mkOption {
+            type = with types; nullOr lines;
+            default = null;
+            description =
+              ''
+                Lua code inserted at the end of the plugin's configuration.
+              ''
+              + lib.optionalString hasContent ''
+                This is the same as using `lib.nixvim.utils.mkAfterSection` when defining `content`.
+              '';
+          };
         };
       };
 
-      config.content = lib.mkMerge (
-        lib.optional (config.pre != null) (mkBeforeSection config.pre)
-        ++ lib.optional (config.post != null) (mkAfterSection config.post)
-      );
-    }
-  );
+      contentModule =
+        { config, ... }:
+        {
+          options = {
+            content = lib.mkOption {
+              type = types.lines;
+              default = "";
+              description = ''
+                Configuration of the plugin.
+
+                If `pre` and/or `post` are non-null, they will be merged using the order priorities
+                ${toString (mkBeforeSection null).priority} and ${toString (mkBeforeSection null).priority}
+                respectively.
+              '';
+            };
+          };
+
+          config = {
+            content = lib.mkMerge (
+              lib.optional (config.pre != null) (mkBeforeSection config.pre)
+              ++ lib.optional (config.post != null) (mkAfterSection config.post)
+            );
+          };
+        };
+    in
+    types.submodule ([ commonModule ] ++ (lib.optional hasContent contentModule));
```

Does this work? I didn't think `types.submodule` could accept a list :thinking: 

```suggestion
    types.submoduleWith {
      modules = [ commonModule ] ++ lib.optional hasContent contentModule;
    };
```

---

https://nixos.org/manual/nixos/stable/#sec-option-types-submodule
> This is equivalent to types.submoduleWith { modules = toList o; shorthandOnlyDefinesConfig = true; }

---

> > This is equivalent to `modules = toList o`

Thanks for educating me! :heart:

---

For me the `let in` seems clearer